quantumblacklabs/kedro-viz
# [KED-1650] New graphing algorithm & feature flags

## Description

Improves pipeline visualisation using a new DAG graphing algorithm approach.

![comparison](https://user-images.githubusercontent.com/45365769/83750576-65e69100-a65d-11ea-871c-13c30b269a99.png)

The algorithm is focused on improving the quality and clarity of output for both node layout and edge routing while maintaining  good performance.

Goals include balancing the regularity, legibility, symmetry, spacing and compactness of layout and edges while scaling to graphs with a large number of nodes and edges, for which the existing solution does not handle well enough for this purpose.

These changes are delivered alongside a minimal feature flag system to support toggling of experimental new features such as this one until their general release.

**The feature flag must be enabled in your browser before use and testing (see below).**

## Development notes

As there is currently no GUI for flags, they must be enabled and disabled using parameters in the URL. The status of all flags will persist in the user's local storage config and the console shows a message regarding the status of all flags on load.

<img width="578" alt="Screenshot 2020-06-04 at 12 18 30" src="https://user-images.githubusercontent.com/45365769/83750729-9b8b7a00-a65d-11ea-99b2-016dd386bc81.png">

The flag for the new graphing algorithm is `newgraph`, it is disabled by default, you can use the URL parameters `newgraph=true` or `newgraph=false`.

e.g.`http://localhost:4141/?data=demo&newgraph=true`.

The implementation itself is self contained and includes a selector that connects it to the rest of the system in the same way as the existing solution, such that it works as a drop-in replacement.

It currently depends on the `kiwi.js` solver, though there is potential that this may not eventually be required given some upgrades to the main solver.

## QA notes

The feature must be enabled by the user before testing, as described above. Try the following test data sets. Side-by-side comparison is possible by enabling or disabling the flag:

__Demo__
http://localhost:4141/?data=demo&newgraph=true
http://localhost:4141/?data=demo&newgraph=false

__Random__ (includes layers)
http://localhost:4141/?data=random&newgraph=true
http://localhost:4141/?data=random&newgraph=false

_(see console to get a reusable seed for direct comparison)_

__Animals__
http://localhost:4141/?data=animals&newgraph=true
http://localhost:4141/?data=animals&newgraph=false

__Nodes.json__ (as generated by Kedro)
http://localhost:4141/?newgraph=true
http://localhost:4141/?newgraph=false

It has been tested on both real samples as well as varied pipelines generated by the existing random generator, both ranging so far from around 10 nodes to around 1000, all showing some strong improvement, both when labels are visible and when not.

It has support for and has been tested with pipelines that include layers.

There are also unit tests covering the majority of the changes, though they are not sufficient to test the actual quality of graphing output which must be done manually. For this I have been considering some metrics for more objective evaluation but more exploration is needed.

While the algorithm appears to work well on most pipelines seen so far, wider beta testing is encouraged before the flag is removed for general release.

## Checklist


## Legal notice


- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
